### ISSUE TITLE
git command

### ISSUE THREAD

**Issue by ywangd** [ywangd/stash#44]
I personally didn't have much success in using the git command in shellista. I can `clone` but not able to `pull` after updates to original repo. It would be great to have a git command that perform at least following tasks:
* clone
* pull
* commit and push
* status

Since the git command requires external libraries (dulwich and gittle as I understand), I didn't want to work on it before there is way to manage libraries. Now that @briarfox is working on the `pip` module, maybe it is finally possible to work on git once `pip` is ready.

**Comment by jsbain** [comment]
One problem may be that transistor1 was using a custom gittle and or dulwich, I forget which.  Partly this was to fix some bugs, which were fixed in later versions of dulwich.   I had to further modify one of them for gitview... I also forget why.  When I was implementing gitview, I tried to point dulwich at the latest repo, which had some enhancements, and there were issues, due to some external dependencies (like they assumed multithreading, or some external unix commands).  Or else they had made some changes to the method signatures in a non backwards compatible way, and I didn't feel like changing everything.  

If you had installed dulwich or gittle from their git repos first, then shellista, you might have issues.  Also, pull probably only worked in the shellista-dev branch of the modular shellista.

The other problem is that push and maybe pull only works in a very limited scenario:  fast forward merges.  Merge is tricky.

**Comment by ywangd** [comment]
@jsbain Thanks again for very useful insights on this topic. I had the feeling that pure python git is not totally viable (very restricted). I vaguely remember running into the not compatible method signature issue. I never had any success on pulling new updates from origin repo. 

I think the expectation should be further detailed as:
* clone (this should be easy)
* pull (for only fast-foward changes)
* commit
* push (again fast-forward only)
* switch to other existing branch (not sure how difficult this may be)
* create new branch and push back to origin (this would be very useful but can be skipped if not possible)

These criteria are sketched based on my minimum needs for doing some real development inside Pythonista. You inputs are highly appreciated. 

Is `gitview` able to perform all above tasks? Will it be OK with you if I use some or even large portion of code from `gitview` for the git command? It would be even more fantastic if you could help on this matter. Thanks a lot!

**Comment by jsbain** [comment]
Gitview has all of these features, mostly copied from shellista.  Shellista might be a better starting point, since it is commandline oriented... also gitview code is kind of disorganized.   You should use my version of gittle, as the official one doesn't work with thin packs... 

I can take a shot at git, if no one else is already working on this.

Few questions:   How do we want to handle external dependencies?  
Try to auto install?  Or install using pip? shellista had a module installer which took care of dulwich, gittle, and funky, the three modules we need.
If so, should I install to site-packages or do you have a local packages folder that I can install into?

What about pythonista specific features?  Is stash intended to be usable on non ios devices, for example where `keychain` is not available?

**Comment by ywangd** [comment]
@jsbain Thanks a lot for your inputs and willingness to work on the git issue. 
I didn't know the `git` command would require customized external libraries to work. I initially expected the dependencies would be handled by `pip` command. Now it seems that we need a package manager for GitHub as well before the `git` command can be fully operational. 

My personal answers to your questions are as follows:
* I'd like external dependencies handled by other commands like `pip` or the future GitHub package managing command (please see my latest comment in #39). The `git` command itself should not worry about the details of how external packages are installed. Ideally, it simply asks stash to ensure the existence of all dependencies, something like:
```python
require('dulwich', source='transistor1@GitHub', version='XXX')
```
* I'd say the first choice is to install to system site-packages folder. But a StaSh specific library folder is possible as I am going to add a `lib` folder under stash (#5).
* I never thought about have stash running on non-ios devices. So I'd say its safe to use all iOS features.

Thanks!

**Comment by ywangd** [comment]
@jsbain Thanks for the pull request!
At this stage, lets just assume the external packages are in place for `git` to use. I think its great that the `git` command does not try to install them. If you agree on my suggestion about adding a GitHub package manager (#39), we can open a new ticket and work on it in parallel with the `git` command.

**Comment by jsbain** [comment]
Package manager sounds good.  As a stopgap, I could simply provide a zip of the dependencies, and use zipimporter.

**Comment by ywangd** [comment]
@jsbain 
I agree that pypi package compatible GitHub repo can be handled by `pip` as well. So hopefully @briarfox would find some time to add this feature to the `git` command. A zip importer sounds great atm. Thanks again for your initiative on this issue.

**Comment by ywangd** [comment]
@jsbain git push does not work.
I tried both `git push` and `git push -u user:pass`. There are two errors:
* `urllib2` is not imported 
* keychain is not defined, I found following code and it does not look right to me:
`keychainservice = 'shellista.git.{0}'.format(netloc)`

Could you please have a look when you have time? Thanks!

**Comment by jsbain** [comment]
I added the keychain and irllib2 imports, but still getting another error... I could have sworn I tested this, but will need to look some more.  At the moment, my ipad is slightly out of commission (screen repair results in wifi barely working) so it may take me a little while to get back to this.

**Comment by steljas** [comment]
git push seems to have general functionality today. Are there any outstanding issues needing to be addressed to close this out?
